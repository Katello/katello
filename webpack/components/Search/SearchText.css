.autocomplete-wrapper {
  width: 100%;
  position: relative;
  display: inline-block;
}

.autocomplete-wrapper > div {
  width: 100%;
}

.autocomplete-wrapper input {
  border: 1px solid #cecece;
  padding: 12px 15px;
  font-size: 18px;
  width: 100%;
}

.autocomplete-wrapper input:focus {
  border-color: #0F67FF;
  box-shadow: none;
  outline: none;
}

.autocomplete-wrapper .dropdown {
  width: 100%;
  padding: 0;
  text-align: left;
  max-height: 280px;
  overflow: hidden;
  overflow-y: auto;
  background-color: #ffffff;
  border: 1px solid #0F67FF;
  border-top: none;
}

.autocomplete-wrapper .item  {
  display: block;
  cursor: pointer;
  font-size: 18;
  padding: 5px;
}

.autocomplete-wrapper .item.selected-item {
  background-color: #0069ff;
  color: #FAFBFC;
}

.autocomplete-wrapper .item:hover {
  background-color: #0069ff;
  color: #FAFBFC;
}
