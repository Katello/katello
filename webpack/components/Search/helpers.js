// eslint-disable-next-line import/prefer-default-export
export const stringIncludes = (string, includes) => {
  const a = string.replace(/\s/g, '').toLowerCase();
  const b = includes.replace(/\s/g, '').toLowerCase();
  return a.includes(b);
};
