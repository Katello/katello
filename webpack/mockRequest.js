import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import nock from 'nock';

// TODO: figure out way to reuse this from foreman
export const mock = new MockAdapter(axios);
const methods = {
  GET: 'onGet',
  POST: 'onPost',
  PUT: 'onPut',
  DELETE: 'onDelete',
};

const errorResponse = msg => ({ displayMessage: msg });

export const mockRequest = ({
  method = 'GET',
  url,
  data,
  status = 200,
  response = null,
}) => mock[methods[method.toUpperCase()]](url, data).reply(status, response);

export const mockErrorRequest = ({
  status = 500,
  ...options
}) => mockRequest({
  response: errorResponse(`Request failed with status code ${status}`),
  status,
  ...options,
});

export const mockReset = () => mock.reset();

// Using the library 'nock' as it matches actual network requests rather than mock another
// library. This is helpful when the request is not coming from Katello. For example, axios
// called within Katello can be mocked with axios-mock-adapter or similar, but a http request
// made by axios that is coming from Foreman cannot be mocked by axios-mock-adapter or a
// jest mock within Katello. So to do this, we can mock the request a level deeper within
// nodejs by using nock.
export const nockInstance = nock('http://localhost');

// Calling .done() with nock asserts that the request was fufilled. We use a timeout to ensure
// that the component has set up and made the request before the assertion is made.
export const assertNockRequest = (scope, timeout = 2000) => {
  setTimeout(() => {
    scope.done();
  }, timeout);
};
