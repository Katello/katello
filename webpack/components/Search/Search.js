import React, { useState, useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { ControlLabel } from 'react-bootstrap';
import { loadSetting } from 'foremanReact/components/Settings/SettingsActions';
import { translate as __ } from 'foremanReact/common/I18n';
import PropTypes from 'prop-types';
import TypeAhead from '../TypeAhead';
import api, { foremanApi } from '../../services/api';
import { stringIncludes } from './helpers';
import {
  AUTOSEARCH_DELAY,
  AUTOSEARCH_WHILE_TYPING,
} from '../../scenes/Settings/SettingsConstants';

const Search = ({
  onSearch,
  updateSearchQuery,
  isDisabled,
  settings: { autoSearchDelay },
  initialInputValue,
  patternfly4,
  settings: { autoSearchEnabled },
  getAutoCompleteParams,
  foremanApiAutoComplete,
  bookmarkController,
}) => {
  const [items, setItems] = useState([]);
  const dispatch = useDispatch();
  const mountedRef = useRef(true);

  const onInputUpdate = async (searchTerm = '') => {
    const newItems = items.filter(({ text }) => stringIncludes(text, searchTerm));

    if (newItems.length !== items.length) {
      // Checking whether the current component is mounted before state change events
      if (!mountedRef.current) return;
      setItems(newItems);
    }

    const { params, endpoint } = getAutoCompleteParams(searchTerm);

    if (endpoint) {
      let data;
      if (foremanApiAutoComplete) {
        data = await foremanApi.get(endpoint, undefined, params);
      } else {
        data = await api.get(endpoint, undefined, params);
      }
      // Checking whether the current component is mounted before state change events
      if (!mountedRef.current) return;
      setItems(data?.data?.filter(({ error }) => !error).map(({ label }) => ({
        text: label.trim(),
      })));
    }

    if (autoSearchEnabled && patternfly4) {
      onSearch(searchTerm || '');
    }
  };

  useEffect(() => {
    dispatch(loadSetting(AUTOSEARCH_DELAY));
    dispatch(loadSetting(AUTOSEARCH_WHILE_TYPING));
    return () => { mountedRef.current = false; };
  }, [dispatch]);

  const onNewSearch = (search) => {
    if (updateSearchQuery) updateSearchQuery(search);
    onSearch(search);
  };

  return (
    <div>
      <ControlLabel srOnly>{__('Search')}</ControlLabel>
      <TypeAhead
        autoSearchDelay={autoSearchDelay}
        bookmarkController={bookmarkController}
        isDisabled={isDisabled}
        items={items}
        onInputUpdate={onInputUpdate}
        onSearch={onNewSearch}
        initialInputValue={initialInputValue}
        patternfly4={patternfly4}
        autoSearchEnabled={autoSearchEnabled}
      />
    </div>
  );
};

Search.propTypes = {
  onSearch: PropTypes.func.isRequired,
  getAutoCompleteParams: PropTypes.func.isRequired,
  foremanApiAutoComplete: PropTypes.bool,
  updateSearchQuery: PropTypes.func,
  initialInputValue: PropTypes.string,
  patternfly4: PropTypes.bool,
  isDisabled: PropTypes.bool,
  settings: PropTypes.shape({
    autoSearchEnabled: PropTypes.bool,
    autoSearchDelay: PropTypes.number,
  }),
  bookmarkController: PropTypes.string,
};

Search.defaultProps = {
  updateSearchQuery: undefined,
  foremanApiAutoComplete: false,
  initialInputValue: '',
  patternfly4: false,
  settings: {
    autoSearchEnabled: true,
  },
  isDisabled: undefined,
  bookmarkController: undefined,
};

export default Search;
