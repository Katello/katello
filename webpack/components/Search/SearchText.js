import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { SearchAutocomplete } from 'foremanReact/components/SearchBar/SearchAutocomplete';
import { useAPI } from 'foremanReact/common/hooks/API/APIHooks';
import { STATUS } from 'foremanReact/constants';
import { noop } from 'foremanReact/common/helpers';

const SearchText = ({
  data: {
    autocomplete: { url, apiParams } = { url: '' },
    disabled,
  },
  initialQuery,
  onSearchChange,
  name,
}) => {
  const [search, setSearch] = useState(initialQuery || '');
  const getAPIparams = input => ({ ...apiParams(input) });
  const { response, status, setAPIOptions } = useAPI('get', url, {
    params: getAPIparams(search),
  });
  const onChange = (newValue) => {
    onSearchChange(newValue);
    setSearch(newValue);
    setAPIOptions({ params: { ...getAPIparams(newValue) } });
  };
  const error =
    status === STATUS.ERROR || response?.[0]?.error
      ? response?.[0]?.error || response.message
      : null;

  let results = [];
  if (Array.isArray(response) && !error) {
    results = response.map(item => ({ label: item, category: '' }));
  }

  return (
    <div className="foreman-search-text">
      <SearchAutocomplete
        results={results}
        onSearchChange={onChange}
        value={search}
        disabled={disabled}
        error={error}
        name={name}
      />
    </div>
  );
};

SearchText.propTypes = {
  data: PropTypes.shape({
    autocomplete: PropTypes.shape({
      url: PropTypes.string.isRequired,
      apiParams: PropTypes.func,
    }).isRequired,
    disabled: PropTypes.bool,
  }).isRequired,
  initialQuery: PropTypes.string,
  onSearchChange: PropTypes.func,
  name: PropTypes.string,
};

SearchText.defaultProps = {
  initialQuery: '',
  onSearchChange: noop,
  name: null,
};

export default SearchText;
