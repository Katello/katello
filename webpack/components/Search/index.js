import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as settingActions from 'foremanReact/components/Settings/SettingsActions';
import { selectSettings } from '../../scenes/Settings/SettingsSelectors';
import Search from './Search';

const mapStateToProps = state => ({
  settings: selectSettings(state),
});

const actions = { ...settingActions };

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Search);
