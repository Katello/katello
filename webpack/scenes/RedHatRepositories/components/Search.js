/* eslint-disable import/no-extraneous-dependencies */
import React, { Component } from 'react';
import { DropdownButton, MenuItem } from 'patternfly-react';
import PropTypes from 'prop-types';

import '../index.scss';
import Search from '../../../components/Search/index';
import { orgId } from '../../../services/api';

class RepositorySearch extends Component {
  constructor(props) {
    super(props);
    this.dropDownItems = [
      {
        key: 'available',
        endpoint: 'repository_sets',
        title: 'Available',
      },
      {
        key: 'enabled',
        endpoint: 'enabled_repositories',
        title: 'Enabled',
      },
      {
        key: 'both',
        endpoint: false,
        title: 'Both',
      },
    ];
    this.state = { searchList: this.dropDownItems[0] };
    this.onSearch = this.onSearch.bind(this);
    this.getAutoCompleteParams = this.getAutoCompleteParams.bind(this);
  }

  onSearch(search) {
    this.props.onSearch(search);
  }

  onSelectSearchList(searchList) {
    this.setState({ searchList });
    this.props.onSelectSearchList(searchList.key);
  }

  getAutoCompleteParams(search) {
    const params = {
      organization_id: orgId(),
      search,
    };

    if (this.state.searchList.key === 'enabled') {
      params.enabled = true;
    }

    return {
      endpoint: '/repository_sets/auto_complete_search',
      params,
    };
  }

  render() {
    return (
      <div style={{ display: 'flex' }}>
        <DropdownButton title={this.state.searchList.title} id="search-list-select">
          {this.dropDownItems
            .filter(({ key }) => key !== this.state.searchList.key)
            .map(({ key, title, ...rest }) => (
              <MenuItem
                key={key}
                onClick={() =>
                  this.onSelectSearchList({ key, title, ...rest })
                }
              >
                {title}
              </MenuItem>
            ))}
        </DropdownButton>
        <Search
          onSearch={this.onSearch}
          getAutoCompleteParams={this.getAutoCompleteParams}
        />
      </div>
    );
  }
}

RepositorySearch.propTypes = {
  onSearch: PropTypes.func.isRequired,
  onSelectSearchList: PropTypes.func.isRequired,
};

export default RepositorySearch;
