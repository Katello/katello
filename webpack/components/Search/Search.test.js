import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { mock as mockApi } from '../../mockRequest';

import Search from '../Search/Search';

describe('Search component', () => {
  const getBaseProps = () => ({
    onSearch: () => {},
    getAutoCompleteParams: () => ({ endpoint: '/fake' }),
    loadSetting: jest.fn(),
  });

  describe('rendering', () => {
    it('renders correctly', () => {
      mockApi.onGet('/katello/api/v2/fake').reply(200, []);
      const component = shallow(<Search {...getBaseProps()} />);

      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
