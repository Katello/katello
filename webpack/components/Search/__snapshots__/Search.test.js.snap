// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component rendering renders correctly 1`] = `
<div>
  <ControlLabel
    bsClass="control-label"
    srOnly={true}
  >
    Search
  </ControlLabel>
  <TypeAhead
    actionText="Search"
    autoSearchEnabled={true}
    initialInputValue=""
    items={Array []}
    onInputUpdate={[Function]}
    onSearch={[Function]}
    patternfly4={false}
  />
</div>
`;
